aiida_pseudo/groups/family/pseudodojo.py

```diff
@@ -0,0 +1,227 @@
+
+    The `PseudoDojoPsp8Family` is essentially a `PseudoDojoFamily` with some additional constraints.
+    It can only be used to contain the pseudo potentials and corresponding metadata of an official psp8
+    PseudoDojo configuration.
+    """
+
+    default_configuration = PseudoDojoConfiguration('04', 'pbe', 'sr', 'standard', 'normal', 'psp8')
+    valid_configurations = (
+        # nc-sr-03_pbe
+        PseudoDojoConfiguration('03', 'pbe', 'sr', 'standard', 'none', 'psp8'),
+        PseudoDojoConfiguration('03', 'pbe', 'sr', 'stringent', 'none', 'psp8'),
+        # nc-sr-03_pbesol
+        PseudoDojoConfiguration('03', 'pbesol', 'sr', 'standard', 'none', 'psp8'),
+        PseudoDojoConfiguration('03', 'pbesol', 'sr', 'stringent', 'none', 'psp8'),
+        # nc-sr-03_lda
+        PseudoDojoConfiguration('03', 'lda', 'sr', 'standard', 'none', 'psp8'),
+        PseudoDojoConfiguration('03', 'lda', 'sr', 'stringent', 'none', 'psp8'),
+        # nc-sr-04_pbe_standard
+        PseudoDojoConfiguration('04', 'pbe', 'sr', 'standard', 'low', 'psp8'),
+        PseudoDojoConfiguration('04', 'pbe', 'sr', 'standard', 'normal', 'psp8'),
+        PseudoDojoConfiguration('04', 'pbe', 'sr', 'standard', 'high', 'psp8'),
+        PseudoDojoConfiguration('04', 'pbe', 'sr', 'standard', 'none', 'psp8'),
+        # nc-sr-04_pbe_stringent
+        PseudoDojoConfiguration('04', 'pbe', 'sr', 'stringent', 'low', 'psp8'),
+        PseudoDojoConfiguration('04', 'pbe', 'sr', 'stringent', 'normal', 'psp8'),
+        PseudoDojoConfiguration('04', 'pbe', 'sr', 'stringent', 'high', 'psp8'),
+        PseudoDojoConfiguration('04', 'pbe', 'sr', 'stringent', 'none', 'psp8'),
+        # nc-sr-04_pbesol_standard
+        PseudoDojoConfiguration('04', 'pbesol', 'sr', 'standard', 'low', 'psp8'),
+        PseudoDojoConfiguration('04', 'pbesol', 'sr', 'standard', 'normal', 'psp8'),
+        PseudoDojoConfiguration('04', 'pbesol', 'sr', 'standard', 'high', 'psp8'),
+        PseudoDojoConfiguration('04', 'pbesol', 'sr', 'standard', 'none', 'psp8'),
+        # nc-sr-04_pbesol_stringent
+        PseudoDojoConfiguration('04', 'pbesol', 'sr', 'stringent', 'low', 'psp8'),
+        PseudoDojoConfiguration('04', 'pbesol', 'sr', 'stringent', 'normal', 'psp8'),
+        PseudoDojoConfiguration('04', 'pbesol', 'sr', 'stringent', 'high', 'psp8'),
+        PseudoDojoConfiguration('04', 'pbesol', 'sr', 'stringent', 'none', 'psp8'),
+        # nc-sr-04_lda
+        PseudoDojoConfiguration('04', 'lda', 'sr', 'standard', 'none', 'psp8'),
+        PseudoDojoConfiguration('04', 'lda', 'sr', 'stringent', 'none', 'psp8'),
+        # nc-sr-04-3plus_pbe
+        PseudoDojoConfiguration('04', 'pbe', 'sr3plus', 'standard', 'none', 'psp8'),
+        # nc-fr-04_pbe_standard
+        # PseudoDojoConfiguration('04', 'pbe', 'fr','standard', 'low', 'psp8'),
+        # PseudoDojoConfiguration('04', 'pbe', 'fr','standard', 'normal', 'psp8'),
```

Comment:
Are all these configurations really different for each family simply based on the format of the underlying pseudo? The format is merely a way of encoding the definition of the pseudo is it not? So then it is technically possible to take a pseudo in one format and transform it to any other format. If that is so, then the same configurations could exist for each of the three formats, could they not?

Reply:
They can, yes. I did this because the name of the family is generated directly from the `PseudoDojoConfiguration`, and I didn't want to have conflicting names when the same configuration with a different file type is present, which would make it inconvenient to load the correct family by label.

Reply:
I don't think I really answered your question the first time.

In theory, one could create files in all the formats with identical pseudopotentials, in which case you would only need one configuration which would apply to all formats.

However, like I said above, some configurations are _currently_ only available for certain subsets of file formats, so I do this to make sure that all the configurations are actually valid.

Reply:
> They can, yes. I did this because the name of the family is generated directly from the PseudoDojoConfiguration, and I didn't want to have conflicting names when the same configuration with a different file type is present, which would make it inconvenient to load the correct family by label.

I see, but this can be solved in a more elegant way. You could retrieve it directly from the pseudo type of the family:
```
    @classmethod
    def format_configuration_label(cls, configuration: PseudoDojoConfiguration) -> str:
        """Format a label for an `PseudoDojoFamily` with the required syntax.

        :param configuration: the PseudoDojo configuration
        :return: label
        """
        from aiida.plugins.entry_point import get_entry_point_from_class

        pseudo_type = cls._pseudo_type
        entry_point_group, entry_point = get_entry_point_from_class(pseudo_type.__module__, pseudo_type.__name__)
        pseudo_format = entry_point.name.split('.')[-1]

        return cls.label_template.format(
            version=configuration.version,
            functional=configuration.functional,
            relativistic=configuration.relativistic,
            protocol=configuration.protocol,
            format=pseudo_format
        )
```

Given that the format does not affect the cutoffs, this also makes a lot more sense semantically speaking. This way you can get rid of all the duplication of specifying the same valid configurations over and over again.

The only difference I can see is that version 0.3 is not available for the Psml families. It would be ideal if those could also be generated of course. If this is really not possible, I would instead just invert the logic a bit and define an `unavailable_configuration` just for the `PseudoDojoPsmlFamily` family.

Reply:
Seems like a reasonable solution. At the beginning, I was hoping that we could keep everything as similar as possible with the SSSP, but at this point that's clearly out the window.

So, file format will be used by the CLI to select the appropriate `PseudoDojoFamily` subclass, and the labelling will be handled by the subclass?

Reply:
> Seems like a reasonable solution. At the beginning, I was hoping that we could keep everything as similar as possible with the SSSP, but at this point that's clearly out the window.

Well this is a minor difference, and only really in the implementation of `get_valid_labels`. The interface remains the same. You can have a look at the commit I pushed.

> So, file format will be used by the CLI to select the appropriate `PseudoDojoFamily` subclass, and the labelling will be handled by the subclass?

The label formatting is still implemented by the `PseudoDojoFamily` class. The CLI will still call this, but by calling it on one of the subclasses, the format part of the label will be automatically set.